dev_handler/Makefile

```diff
@@ -1,10 +1,30 @@
-LIBFLAGS=-pthread -lrt -Wall
-BIN=../bin
+# list of libraries that dev_handler needs to compile
+LIBS=-pthread -lrt -Wall
 
-dev_handler: dev_handler.c message.c message.h ../logger/logger.c ../runtime_util/runtime_util.c ../shm_wrapper/shm_wrapper.c
-	gcc $^ -o $(BIN)/dev_handler $(LIBFLAGS)
+# list of source files that the target (dev_handler) depends on, relative to this folder
+SRCS = dev_handler.c dev_handler_message.c ../logger/logger.c ../runtime_util/runtime_util.c ../shm_wrapper/shm_wrapper.c
 
+# specify the target (executable we want to make)
+TARGET = dev_handler
+
+.PHONY: all clean $(TARGET)
+
+all: $(TARGET)
+
+# include top-level Makefile for some common variables and rules
+include ../Makefile_defs
+
+# resolve phony target "dev_handler" as ../bin/dev_handler
+$(TARGET): $(BIN)/$(TARGET)
+
+# rule to compile dev_handler
+$(BIN)/$(TARGET): $(OBJS) | $(BIN)
+	$(CC) $(OBJS) -o $@ $(LIBS)
+
+# remove build artifacts
 clean:
-	rm -f $(BIN)/dev_handler
-	rm -f dev_handler
-    
+	rm -f $(OBJS)
+	rm -f $(patsubst %.o,%.d,$(OBJS))
+	rm -f $(BIN)/$(TARGET)
+
+-include $(patsubst %.o,%.d,$(OBJS))
```

Comment:
What's this do?

Reply:
This is a little complex. So basically the `-MMD -MP` compiler flags that I added to `CFLAGS` at the top of `Makefile_defs` tells the compiler that, when compiling a source file into an object file, to also create a "dependency file" for that object file. The dependency file is in the format of a Makefile and basically lists out all of the sources and header files that the compiler needed to look at in order to generate the object file. The dependency file has the same name as the generated object file, except with a `.d` extension. So we include all of the `.d` dependency files associated with the object files needed to make this executable with this line. Thus, when we try to compile a specific object file, `make` will know which sources and headers that object file is dependent on, and if any of those are newer than the object file we are trying to make, then the object file will be recompiled.

Here are a couple of good articles explaining it: 

https://nathandumont.com/blog/automatically-detect-changes-in-header-files-in-a
https://stackoverflow.com/questions/8025766/makefile-auto-dependency-generation
http://www.microhowto.info/howto/automatically_generate_makefile_dependencies.html